protocols/upnp/src/behaviour.rs

```diff
@@ -0,0 +1,515 @@
+    NetworkBehaviour, NewListenAddr, PollParameters, ToSwarm,
+};
+
+/// The duration in seconds of a port mapping on the gateway.
+const MAPPING_DURATION: u64 = 3600;
+
+/// Renew the Mapping every half of `MAPPING_DURATION` to avoid the port being unmapped.
+const MAPPING_TIMEOUT: u64 = MAPPING_DURATION / 2;
+
+/// Map a port on the gateway.
+fn map_port<P: Provider + 'static>(
+    gateway: Arc<P::Gateway>,
+    mapping: Mapping,
+    permanent: bool,
+) -> BoxFuture<'static, Event> {
+    let duration = if permanent { 0 } else { MAPPING_DURATION };
+
+    P::add_port(
+        gateway,
+        mapping.protocol,
+        mapping.internal_addr,
+        Duration::from_secs(duration),
+    )
+    .map(move |result| match result {
+        Ok(()) => Event::Mapped(mapping),
+        Err(err) => Event::MapFailure(mapping, err),
+    })
+    .boxed()
+}
+
+/// Remove a port mapping on the gateway.
+fn remove_port_mapping<P: Provider + 'static>(
+    gateway: Arc<P::Gateway>,
+    mapping: Mapping,
+) -> BoxFuture<'static, Event> {
+    P::remove_port(gateway, mapping.protocol, mapping.internal_addr.port())
+        .map(move |result| match result {
+            Ok(()) => Event::Removed(mapping),
+            Err(err) => Event::RemovalFailure(mapping, err),
+        })
+        .boxed()
+}
+
+/// A [`Provider::Gateway`] event.
+#[derive(Debug)]
+enum Event {
+    /// Port was successfully mapped.
+    Mapped(Mapping),
+    /// There was a failure mapping port.
+    MapFailure(Mapping, Box<dyn Error>),
+    /// Port was successfully removed.
+    Removed(Mapping),
+    /// There was a failure removing the mapping port.
+    RemovalFailure(Mapping, Box<dyn Error>),
+}
+
+/// Mapping of a Protocol and Port on the gateway.
+#[derive(Debug, Clone)]
+struct Mapping {
+    listener_id: ListenerId,
+    protocol: Protocol,
+    multiaddr: Multiaddr,
+    internal_addr: SocketAddrV4,
+}
+
+impl Hash for Mapping {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.listener_id.hash(state);
+    }
+}
+
+impl PartialEq for Mapping {
+    fn eq(&self, other: &Self) -> bool {
+        self.listener_id == other.listener_id
+    }
+}
+
+impl Eq for Mapping {}
+
+impl Borrow<ListenerId> for Mapping {
+    fn borrow(&self) -> &ListenerId {
+        &self.listener_id
+    }
+}
+
+/// Current state of a [`Mapping`].
+enum MappingState {
+    /// Port mapping is inactive, will be requested or re-requested on the next iteration.
+    Inactive,
+    /// Port mapping/removal has been requested on the gateway.
+    Pending,
+    /// Port mapping is active with the inner timeout.
+    Active(Delay),
+    /// Port mapping is permanent on the Gateway.
+    Permanent,
+}
+
+/// Current state of the UPnP [`Provider::Gateway`].
+enum GatewayState<P: Provider> {
+    Searching(BoxFuture<'static, Result<(P::Gateway, Ipv4Addr), Box<dyn std::error::Error>>>),
+    Available((Arc<P::Gateway>, Ipv4Addr)),
+    GatewayNotFound,
```

Comment:
Should we periodically try again?

Reply:
> if the gateway doesn't support UPnP it's redundant to try again

That entirely depends on the uptime of the software right? In theory, the hardware could get changed without our software rebooting rebooting, right? I wonder what a good trigger would be to try and search again. Should we monitor the different IPs that get reported via `ExternalAddrCandidate`?

Reply:
Perhaps it is better to not try to be too smart here and require that users reboot their software if they make hardware configuration changes.

Reply:
> > if the gateway doesn't support UPnP it's redundant to try again
> 
> That entirely depends on the uptime of the software right? In theory, the hardware could get changed without our software rebooting rebooting, right? I wonder what a good trigger would be to try and search again. Should we monitor the different IPs that get reported via `ExternalAddrCandidate`?

The software could technically try again, probably in shorter intervals if the initial command failed because it doesnt support upnp. The other option might be to check interface changes. Possibly if libp2p report to the behaviour about a change in the listener that could also be another trigger to try again too.

Reply:
Reacting on `FromSwarm::NewListener`, i.e. each time there is an interface change, sounds worth exploring. That said, fine with the first iteration to just stop. I just suggest logging on debug in such case.